File: android/app/src/main/java/info/bagen/dwebbrowser/microService/helper/StateObservable.kt
package info.bagen.dwebbrowser.microService.helper

import org.dweb_browser.microservice.ipc.Ipc
import org.dweb_browser.microservice.ipc.message.IpcEvent
import info.bagen.dwebbrowser.util.IsChange
import org.dweb_browser.helper.*

open class StateObservable(
    needFirstCall: Boolean = false,
    private val getStateJson: () -> String,
) {
    val stateChanges = IsChange(needFirstCall)
    val changeSignal = SimpleSignal()
    inline fun observe(noinline cb: SimpleCallback) = changeSignal.listen(cb)

    val observerIpcMap = mutableMapOf<Ipc, OffListener>()
    suspend fun startObserve(ipc: Ipc) {
        observerIpcMap.getOrPut(ipc) {
            observe {
                ipc.postMessage(
                    IpcEvent.fromUtf8(
                        "observe",
                        getStateJson()
                    )
                )
            }
        }
    }

    inline fun notifyObserver() {
        runBlockingCatching {
          changeSignal.emit()
        }.getOrNull()
    }

    fun stopObserve(ipc: Ipc) = observerIpcMap.remove(ipc)?.let { off ->
        off(Unit)
        true
    } ?: false
}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/NativeUI/Helper/StateObservable.cs
using System.Collections;

#nullable enable

namespace DwebBrowser.Helper;

public class StateObservable<T>
{
    private Func<string> _getStateJson;

    public StateObservable(Func<string> getStateJson)
    {
        _getStateJson = getStateJson;
    }

    private Dictionary<Ipc, Signal> _observeIpcMap = new();

    private event Signal? _onChange;
    public Task EmitAsync() => (_onChange?.Emit()).ForAwait();

    public Unit StartObserve(Ipc ipc)
    {
        _observeIpcMap.GetValueOrPut(ipc, () =>
        {
            Signal onChange = async (_) =>
            {
                await ipc.PostMessageAsync(IpcEvent.FromUtf8("observe", _getStateJson()));
            };

            _onChange += onChange;
            return onChange;
        });

        return unit;
    }

    public bool StopObserve(Ipc ipc)
    {
        if (_observeIpcMap.Remove(ipc, out var onChange))
        {
            _onChange -= onChange;
            return true;
        }

        return false;
    }
}
